### PR TITLE
[RFC FS-1093] Additional type directed conversions

### PR DESCRIPTION
**Pull request by dsyme**
Implementation of RFC https://github.com/fsharp/fslang-design/blob/master/RFCs/FS-1093-additional-conversions.md

This primarily loosens the auto-widening rules in the presence of explicit type annotations or other "known" type information, e.g. to allow things like:

    let x: obj = if true then 1 else 3.0

It also adds `int32 --> int64` conversions and a limited form of `op_Implicit` support.

In this PR the rule is:

- When processing an expression the "overall type" is usually labelled as "MustConvertTo" the known type, instead of "MustEqual"
 
- This is propagated over if/then/else and other control structures.  So if `if then A then B` must convert to T, then both A and B must convert to T, but A and B may have different types.

- For leaf expressions, we often have immediate knowledge of both the source and MustConvertTo types.  If a unification between the two would have failed then subsumption and adhoc conversions can be tried instead.

The adhoc conversions enabled are in the RFC.

This also allows (the removal of upcast coercions like `asExpr` on this line](https://github.com/dotnet/fsharp/blob/main/src/fsharp/FSharp.Core/Linq.fs#L282), e.g. in favour of one explicit type annotation.

A longer list of things enabled is below and in the RFC

```fsharp

let x1 : obj = (1,2)
let x2 : obj * obj = (1,2)
let x3 : (obj * obj) = (let x = 1 in let y = 2 in (x,y))
let x4 : struct (obj * obj) = struct (1,2)
let x5 : (unit -> obj) = (fun () -> 1)
let x6 : (unit -> obj * obj) = (fun () -> (1,2))
let x7 () : obj * obj = (1,2)
let x8 (s:string) : obj = s

type R = { mutable F1: (obj * obj) }

let annotations = 
    let _ : obj = (1,2)
    let _ : obj * obj = (1,2)

    // structure through let
    let _ : (obj * obj) = (let x = 1 in let y = 2 in (x,y))

    // structure through let rec
    let _ : (obj * obj) = (let rec f x = x in (3,4.0))

    // structure through sequence
    let _ : (obj * obj) = (); (3,4.0)

    // struct tuple
    let _ : struct (obj * obj) = struct (1,2)
    
    // record (both field and overall result)
    // let _ : obj = { F1 = (1, 2) } // TODO
    
    // record (both field and overall result)
    { F1 = (1, 2uy) }.F1 <- (3.0, 4)
    
    // anon record
    let _ : {| A: obj |} = {| A = 1 |}
    
    // lambda return
    let _ : (unit -> obj) = (fun () -> 1)

    // function lambda return
    let _ : (int -> obj) = (function 1 -> 1 | 2 -> 3.0 | _ -> 4uy)

    let _ : (unit -> obj * obj) = (fun () -> (1,2))

    // constants
    (1 :> System.IComparable) |> ignore

    let _ : System.IComparable = 1 

    // array constants
    let _ : System.Array = [| 1us |] 

    let _ : System.Array = [| 1I |] 

    let _ : System.IComparable = 1I
    
    // property
    let _ : System.IComparable<string> = System.String.Empty 
    
    // method
    let _ : System.IComparable<string> = System.String.Format("")

    let _ : obj = System.String.Format("")

    let _ : System.IComparable = System.String.Format("")
    
    // array constants

    let _ : obj[] = [| 1 |] 
    let _ : (obj * obj)[] = [| (1,1) |]
    let _ : (obj * obj)[] = [| ("abc",1) |]
    let _ : (string * obj)[] = [| ("abc",1) |]
    let _ : (string * obj)[] = [| ("abc",1); ("abc",3.0) |]
    let _ : (string * obj)[] = [| Unchecked.defaultof<_>; ("abc",3.0) |]
    let _ : struct (string * obj)[] = [| Unchecked.defaultof<_>; struct ("abc",3.0) |]

    let _ : obj = 1
    let _ : obj = ""
    let _ : obj = { new System.ICloneable with member x.Clone() = obj() }
    let _ : obj = ""
    let _ : obj = string ""
    let _ : obj = id ""
    
    // conditional
    let _ : obj = if true then 1 else 3.0
    let _ : obj = (if true then 1 else 3.0)
    let _ : obj = (if true then 1 elif true then 2uy else 3.0) 
    
    // try-with
    let _ : obj = try 1 with _ -> 3.0
    
    // try-finally
    let _ : obj = try 1 finally ()

    // match
    let _ : obj = match true with _ -> 1 | _ -> 3.0
    ()

let f1 () : obj = 1
let f2 () : obj = if true then 1 else 3.0
```

### CODE REVIEW

**Comment by dsyme** [comment]
Test failures (all expected, at a first glance)

```
2021-01-16T17:04:27.7873329Z Conformance\Expressions\Type-relatedExpressions (E_rigidtypeannotation02.fs) -- failed
2021-01-16T17:04:27.7904419Z Conformance\Expressions\Type-relatedExpressions (E_rigidtypeannotation02b.fs) -- failed
2021-01-16T17:04:30.5798506Z Conformance\InferenceProcedures\ConstraintSolving (E_NoImplicitDowncast01.fs) -- failed
2021-01-16T17:06:47.8831078Z Conformance\ObjectOrientedTypeDefinitions\ClassTypes\Misc (E_ExplicitConstructor.fs) -- failed
2021-01-16T17:09:35.2129707Z Conformance\TypesAndTypeConstraints\CheckingSyntacticTypes (DefaultConstructorConstraint02.fs) -- failed
```

**Comment by charlesroddie** [comment]
This would implement https://github.com/fsharp/fslang-suggestions/issues/849

**Comment by dsyme** [comment]
I set up auto-flow from `main` to this branch here: https://github.com/dotnet/roslyn-tools/blob/master/src/GitHubCreateMergePRs/config.xml

**Comment by dsyme** [comment]
RFC is here: https://github.com/fsharp/fslang-design/blob/master/RFCs/FS-1093-additional-conversions.md

**Comment by Swoorup** [comment]
Testing this branch along with anonymous union.
Looks like records and DUs fail to widen.

```fsharp
type DU = DU of int
let _: obj = DU(0) //error FS0001: This expression was expected to have type ...

type R = { a: int }
let _: obj = { a = 0 } //error FS0001: This expression was expected to have type ...

let _: obj = {|a=1|} // works

[<Measure>]
type id
type UserId = int<id>
let id: UserId = 0<id>
let _: obj = id // works
```

**Comment by dsyme** [comment]
> Testing this branch along with anonymous union. Looks like records and DUs fail to widen.

Thanks, yes, I'll get these fixed - I knew about records but not DUs

**Comment by dsyme** [comment]
> Looks like records and DUs fail to widen.

These are now fixed (subject to CI going green)

**Comment by dsyme** [comment]
The test `BY_DESIGN.ExplicitlyCloseTheParens.Bug73940` is hanging in Windows_vs_release run, I'll disable it and work out why

**Comment by dsyme** [comment]
@brettfo Another one of those test failures
```
The active test run was aborted. Reason: Test host process crashed : Unhandled exception. System.Threading.Tasks.TaskCanceledException: A task was canceled.
   at Microsoft.FSharp.Control.AsyncPrimitives.Start@893-1.Invoke(ExceptionDispatchInfo edi)
   at Microsoft.FSharp.Control.Trampoline.Execute(FSharpFunc`2 firstAction) in /Users/runner/work/1/s/src/fsharp/FSharp.Core/async.fs:line 104
   at Microsoft.FSharp.Control.TrampolineHolder.ExecuteWithTrampoline(FSharpFunc`2 firstAction) in /Users/runner/work/1/s/src/fsharp/FSharp.Core/async.fs:line 169
   at Microsoft.FSharp.Control.AsyncPrimitives.continuation@951-2(AsyncActivation`1 ctxt, Boolean useCcontForTaskCancellation, Task completedTask) in /Users/runner/work/1/s/src/fsharp/FSharp.Core/async.fs:line 952
   at Microsoft.FSharp.Control.AsyncPrimitives.taskContinueWithUnit(Task task, AsyncActivation`1 ctxt, Boolean useCcontForTaskCancellation) in /Users/runner/work/1/s/src/fsharp/FSharp.Core/async.fs:line 966
   at <StartupCode$FSharp-Core>.$Async.AwaitTask@1708-2.Invoke(AsyncActivation`1 ctxt) in /Users/runner/work/1/s/src/fsharp/FSharp.Core/async.fs:line 1708
   at Microsoft.FSharp.Control.AsyncPrimitives.ProtectedCode[T](AsyncActivation`1 ctxt, FSharpFunc`2 userCode) in /Users/runner/work/1/s/src/fsharp/FSharp.Core/async.fs:line 448
   at <StartupCode$FSharp-Core>.$Async.OnCancel@1690-5.Invoke(AsyncActivation`1 ctxt) in /Users/runner/work/1/s/src/fsharp/FSharp.Core/async.fs:line 1690
   at Microsoft.FSharp.Control.Trampoline.Execute(FSharpFunc`2 firstAction) in /Users/runner/work/1/s/src/fsharp/FSharp.Core/async.fs:line 104
   at Microsoft.FSharp.Control.TrampolineHolder.ExecuteWithTrampoline(FSharpFunc`2 firstAction) in /Users/runner/work/1/s/src/fsharp/FSharp.Core/async.fs:line 169
   at <StartupCode$FSharp-Core>.$Async.-ctor@155-1.Invoke(Object o) in /Users/runner/work/1/s/src/fsharp/FSharp.Core/async.fs:line 157
   at System.Threading.QueueUserWorkItemCallbackDefaultContext.Execute()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
```

**Comment by brettfo** [comment]
@dsyme Do you know which test caused this?  The stack is all in the async handler.

**Comment by dsyme** [comment]
@brettfo 

> Do you know which test caused this? The stack is all in the async handler

OnCancel is on the stack so I suspect one of those three tests we identified, unless they're already disabled.

We should also run the tests using --blame or similar on a PR?

**Comment by brettfo** [comment]
Turns out my PR to skip them got stuck: #11587  Working on un-sticking it now.

**Comment by dsyme** [comment]
@brettfo I got a test failure again on MacOS - we were expecting these gone, right? 

I think we should set up a PR running tests on MacOS with `--blame` so we can identify a specific test?


```
The active test run was aborted. Reason: Test host process crashed : go
Unhandled exception. System.Threading.Tasks.TaskCanceledException: A task was canceled.
   at Microsoft.FSharp.Control.AsyncPrimitives.Start@893-1.Invoke(ExceptionDispatchInfo edi)
   at Microsoft.FSharp.Control.Trampoline.Execute(FSharpFunc`2 firstAction) in /Users/runner/work/1/s/src/fsharp/FSharp.Core/async.fs:line 104
   at Microsoft.FSharp.Control.AsyncPrimitives.continuation@951-2(AsyncActivation`1 ctxt, Boolean useCcontForTaskCancellation, Task completedTask) in /Users/runner/work/1/s/src/fsharp/FSharp.Core/async.fs:line 952
   at Microsoft.FSharp.Control.AsyncPrimitives.taskContinueWithUnit(Task task, AsyncActivation`1 ctxt, Boolean useCcontForTaskCancellation) in /Users/runner/work/1/s/src/fsharp/FSharp.Core/async.fs:line 966
   at <StartupCode$FSharp-Core>.$Async.AwaitTask@1708-2.Invoke(AsyncActivation`1 ctxt) in /Users/runner/work/1/s/src/fsharp/FSharp.Core/async.fs:line 1708
   at Microsoft.FSharp.Control.AsyncPrimitives.ProtectedCode[T](AsyncActivation`1 ctxt, FSharpFunc`2 userCode) in /Users/runner/work/1/s/src/fsharp/FSharp.Core/async.fs:line 453
   at <StartupCode$FSharp-Core>.$Async.OnCancel@1690-5.Invoke(AsyncActivation`1 ctxt) in /Users/runner/work/1/s/src/fsharp/FSharp.Core/async.fs:line 1690
   at Microsoft.FSharp.Control.Trampoline.Execute(FSharpFunc`2 firstAction) in /Users/runner/work/1/s/src/fsharp/FSharp.Core/async.fs:line 104
   at <StartupCode$FSharp-Core>.$Async.-ctor@155-1.Invoke(Object o) in /Users/runner/work/1/s/src/fsharp/FSharp.Core/async.fs:line 157
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
```

File: src/fsharp/FSComp.txt
Comment (baronfel) [inline src/fsharp/FSComp.txt:1555]: for the two new diagnostics, I think it would be helpful to specify _which_ `op_Implicit` is used (ie render the type name through the DisplayEnv so that the exact method name the user would put in) in the message itself. If it's part of the message it becomes very easy to make a codefix that inserts the expression for the user to 'solve' the warning.

File: src/fsharp/FSComp.txt
Comment (baronfel) [inline src/fsharp/FSComp.txt:1556]: And for this ambiguous overload case, doing the now-common full list of potential members would be very useful.  These lists are used in FSAC to power the quick fixes and I at least use them often!

Comment (dsyme) [reply src/fsharp/FSComp.txt:1557]: Yes agreed, thanks!  I'll make that change

File: src/fsharp/FSComp.txt
Comment (dsyme) [reply src/fsharp/FSComp.txt:1556]: Yes agreed, thanks!

**Comment by dsyme** [comment]
Failure after merging `main`, it looks unrelated to this branch and I've seen it in another branch
```
2021-06-24T21:01:53.2616428Z   Failed Regression.NoError.Timestamps.Bug3368b [1 s]
2021-06-24T21:00:18.4256261Z   Failed Regression.ModulesFromExternalLibrariesBug5785 [1 s]
```

**Comment by dsyme** [comment]
Transient failures: 
```

  Stack Trace:
     at FSharp.Compiler.Scripting.DependencyManager.UnitTests.DependencyManagerInteractiveTests.SmokeTest - #r nuget package not found() in /Users/runner/work/1/s/tests/FSharp.Compiler.Private.Scripting.UnitTests/DependencyManagerInteractiveTests.fs:line 59
[xUnit.net 00:02:52.12]     FSharp.Compiler.Scripting.DependencyManager.UnitTests.DependencyManagerInteractiveTests.Use AssemblyResolver to resolve assemblies [FAIL]
  Failed FSharp.Compiler.Scripting.DependencyManager.UnitTests.DependencyManagerInteractiveTests.Use AssemblyResolver to resolve assemblies [31 s]
  Error Message:
   resolve failed
Expected: True
Actual:   False
  Stack Trace:
     at FSharp.Compiler.Scripting.DependencyManager.UnitTests.DependencyManagerInteractiveTests.Use AssemblyResolver to resolve assemblies() in /Users/runner/work/1/s/tests/FSharp.Compiler.Private.Scripting.UnitTests/DependencyManagerInteractiveTests.fs:line 469
PackageManagementError 3217 : PackageManager can not reference the System Package 'FSharp.Core'
[xUnit.net 00:03:25.90]     FSharp.Compiler.Scripting.DependencyManager.UnitTests.DependencyManagerInteractiveTests.Verify that referencing FSharp.Core succeeds with CSharp Scripts [FAIL]
  Failed FSharp.Compiler.Scripting.DependencyManager.UnitTests.DependencyManagerInteractiveTests.Verify that referencing FSharp.Core succeeds with CSharp Scripts [27 s]
  Error Message:
   resolve failed but should have succeeded
Expected: True
Actual:   False
  Stack Trace:
     at FSharp.Compiler.Scripting.DependencyManager.UnitTests.DependencyManagerInteractiveTests.Verify that referencing FSharp.Core succeeds with CSharp Scripts() in /Users/runner/work/1/s/tests/FSharp.Compiler.Private.Scripting.UnitTests/DependencyManagerInteractiveTests.fs:line 552
[xUnit.net 00:04:19.97]     FSharp.Compiler.Scripting.UnitTests.InteractiveTests.ML - use assembly with native dependencies [FAIL]
  Failed FSharp.Compiler.Scripting.UnitTests.InteractiveTests.ML - use assembly with native dependencies [32 s]
  Error Message:
   System.Exception : Evaluation returned 15 errors:
	input.fsx (4,1)-(4,49) interactive error /Users/runner/.dotnet/sdk/5.0.301/NuGet.targets(131,5): error : Failed to retrieve information about 'Microsoft.ML.LightGBM' from remote source 'https://api.nuget.org/v3-flatcontainer/microsoft.ml.lightgbm/index.json'. [/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/nuget/7053--a71090e2-a439-4ece-811f-f268d39e0549/Project.fsproj]
	input.fsx (4,1)-(4,49) interactive error /Users/runner/.dotnet/sdk/5.0.301/NuGet.targets(131,5): error :   Operation timed out (api.nuget.org:443) [/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/nuget/7053--a71090e2-a439-4ece-811f-f268d39e0549/Project.fsproj]
	input.fsx (4,1)-(4,49) interactive error /Users/runner/.dotnet/sdk/5.0.301/NuGet.targets(131,5): error :   Operation timed out [/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/nuget/7053--a71090e2-a439-4ece-811f-f268d39e0549/Project.fsproj]
	input.fsx (9,16)-(9,20) typecheck error The namespace 'Data' is not defined.
	input.fsx (21,19)-(21,28) typecheck error The value, namespace, type or module 'DataFrame' is not defined. Maybe you want one of the following:
   Data
	input.fsx (22,56)-(22,78) typecheck error Lookup on object of indeterminate type based on information prior to this program point. A type annotation may be needed prior to this program point to constrain the type of the object. This may allow the lookup to be resolved.
	input.fsx (23,28)-(23,50) typecheck error Lookup on object of indeterminate type based on information prior to this program point. A type annotation may be needed prior to this program point to constrain the type of the object. This may allow the lookup to be resolved.
	input.fsx (26,21)-(26,44) typecheck error The operator 'expr.[idx]' has been used on an object of indeterminate type based on information prior to this program point. Consider adding further type constraints
	input.fsx (28,16)-(28,18) typecheck error The namespace 'ML' is not defined.
	input.fsx (29,16)-(29,18) typecheck error The namespace 'ML' is not defined.
	input.fsx (30,16)-(30,18) typecheck error The namespace 'ML' is not defined.
	input.fsx (32,17)-(32,26) typecheck error The value or constructor 'MLContext' is not defined. Maybe you want one of the following:
   AppContext
	input.fsx (33,18)-(33,32) typecheck error Lookup on object of indeterminate type based on information prior to this program point. A type annotation may be needed prior to this program point to constrain the type of the object. This may allow the lookup to be resolved.
	input.fsx (34,14)-(34,32) typecheck error Lookup on object of indeterminate type based on information prior to this program point. A type annotation may be needed prior to this program point to constrain the type of the object. This may allow the lookup to be resolved.
	input.fsx (35,15)-(35,32) typecheck error Lookup on object of indeterminate type based on information prior to this program point. A type annotation may be needed prior to this program point to constrain the type of the object. This may allow the lookup to be resolved.
  Stack Trace:
     at FSharp.Test.ScriptHelpers.TestHelpers.getValue(FSharpResult`2 value, FSharpDiagnostic[] errors) in /Users/runner/work/1/s/tests/FSharp.Test.Utilities/ScriptHelpers.fs:line 141
   at FSharp.Compiler.Scripting.UnitTests.InteractiveTests.ML - use assembly with native dependencies() in /Users/runner/work/1/s/tests/FSharp.Compiler.Private.Scripting.UnitTests/FSharpScriptTests.fs:line 240
[xUnit.net 00:04:49.43]     FSharp.Compiler.Scripting.UnitTests.InteractiveTests.Script using System.Configuration succeeds [FAIL]
```

**Comment by dsyme** [comment]
I have updated the error messages.  

@TIHan @KevinRansom @vzarytovskii This is now ready for incorporation into main.

**Comment by Happypig375** [comment]
Would it make sense to leave a warning/error for
```fs
type [<AbstractClass>] A() =
    abstract X : unit -> obj
    default _.X() = obj()
type B() =
    inherit A()
    override _.X() = "" // Here
```
to leave a better syntax for covariant return types?

**Comment by dsyme** [comment]
@brettfo I'm getting XLF-related errors however as far as I can tell my XLF files are up-to-date.  I'm not sure how to unblock this


```
C:\Users\vsagent\.nuget\packages\xlifftasks\1.0.0-beta.21367.1\build\XliffTasks.targets(91,5): error : '..\xlf\FSComp.txt.ja.xlf' is out-of-date with 'D:\workspace\_work\1\s\artifacts\obj\FSharp.Compiler.Service\Release\netstandard2.0\FSComp.resx'. Run `msbuild /t:UpdateXlf` to update .xlf files or set UpdateXlfOnBuild=true to update them on every build, but note that it is strongly discouraged to set UpdateXlfOnBuild=true in official/CI build environments as they should not modify source code during the build. [D:\workspace\_work\1\s\src\fsharp\FSharp.Compiler.Service\FSharp.Compiler.Service.fsproj]
##[error]C:\Users\vsagent\.nuget\packages\xlifftasks\1.0.0-beta.21367.1\build\XliffTasks.targets(91,5): error : (NETCORE_ENGINEERING_TELEMETRY=Build) '..\xlf\FSComp.txt.ja.xlf' is out-of-date with 'D:\workspace\_work\1\s\artifacts\obj\FSharp.Compiler.Service\Release\netstandard2.0\FSComp.resx'. Run `msbuild /t:UpdateXlf` to update .xlf files or set UpdateXlfOnBuild=true to update them on every build, but note that it is strongly discouraged to set UpdateXlfOnBuild=true in official/CI build environments as they should not modify source code during the build.
C:\Users\vsagent\.nuget\packages\xlifftasks\1.0.0-beta.21367.1\build\XliffTasks.targets(91,5): error : '..\xlf\FSComp.txt.ja.xlf' is out-of-date with 'D:\workspace\_work\1\s\artifacts\obj\FSharp.Compiler.Service\Release\net472\FSComp.resx'. Run `msbuild /t:UpdateXlf` to update .xlf files or set UpdateXlfOnBuild=true to update them on every build, but note that it is strongly discouraged to set UpdateXlfOnBuild=true in official/CI build environments as they should not modify source code during the build. [D:\workspace\_work\1\s\src\fsharp\FSharp.Compiler.Service\FSharp.Compiler.Service.fsproj]
##[error]C:\Users\vsagent\.nuget\packages\xlifftasks\1.0.0-beta.21367.1\build\XliffTasks.targets(91,5): error : (NETCORE_ENGINEERING_TELEMETRY=Build) '..\xlf\FSComp.txt.ja.xlf' is out-of-date with 'D:\workspace\_work\1\s\artifacts\obj\FSharp.Compiler.Service\Release\net472\FSComp.resx'. Run `msbuild /t:UpdateXlf` to update .xlf files or set UpdateXlfOnBuild=true to update them on every build, but note that it is strongly discouraged to set UpdateXlfOnBuild=true in official/CI build environments as they should not modify source code during the build.
```

**Comment by dsyme** [comment]
@brettfo Oh main is failing to build with the same error https://dev.azure.com/dnceng/public/_build/results?buildId=1256052&view=results

**Comment by dsyme** [comment]
> Would it make sense to leave a warning/error for... to leave a better syntax for covariant return types?

I don't think so.  Covariant return types would I think require a type annotation anyway, as F# uses the abstract slot signature to eagerly lay down type constraints before processing the implementation

**Comment by Happypig375** [comment]
But we can still disable automatic upcasting at override returns (and go on the error path before this PR) and emit a covariant return instead?

**Comment by Happypig375** [comment]
Like
```fs
type [<AbstractClass>] A() =
    abstract A: unit -> obj
type B() =
    inherits A()
    override A() = () // Covariant return
type C() =
    inherits A()
    override A() = id<_> // Infer obj since type inference kicks in and no other type is known
type D() =
    inherits A()
    override A() = () :> _ // Infer obj type for type hole
```

**Comment by dsyme** [comment]
> But we can still disable automatic upcasting at override returns (and go on the error path before this PR) and emit a covariant return instead?

I don't think so, because existing code already relies on the eager application of the non-covariant type information.

To be honest, co-variant returns are just unlikely to ever make it into F#.  Remember F# strongly biases against class/interface hierarchies as a modelling technique (though this stuff does have its uses).  

If we ever add them, I'm pretty certain we would want strong type information for updated types for the slots.

```
type [<AbstractClass>] A() =
    abstract A: unit -> obj
type B() =
    inherits A()
    override A() : string = "a"
```

**Comment by dsyme** [comment]
@KevinRansom This is now ready for preview.

File: tests/FSharp.Compiler.Service.Tests/FSharp.Compiler.Service.Tests.fsproj
Comment (KevinRansom) [inline tests/FSharp.Compiler.Service.Tests/FSharp.Compiler.Service.Tests.fsproj:5]: Probably should go back to :
````
    <TargetFrameworks Condition="'$(OS)' != 'Unix'">net472;net5.0</TargetFrameworks>
    <TargetFrameworks Condition="'$(OS)' == 'Unix'">net5.0</TargetFrameworks>
````

File: tests/FSharp.Test.Utilities/FSharp.Test.Utilities.fsproj
Comment (KevinRansom) [inline tests/FSharp.Test.Utilities/FSharp.Test.Utilities.fsproj:4]: Probably should revert

File: tests/fsharp/tests.fs
Comment (KevinRansom) [inline tests/fsharp/tests.fs:1064]: Why _ ?

**Review by KevinRansom — COMMENTED** [review]
a few things

File: tests/FSharp.Compiler.Service.Tests/FSharp.Compiler.Service.Tests.fsproj
Comment (baronfel) [reply tests/FSharp.Compiler.Service.Tests/FSharp.Compiler.Service.Tests.fsproj:5]: why is this? non-windows boxes can built fr net472 as long as the reference assembly nuget packages are referenced, shouldn't wr move that direction?

Comment (dsyme) [reply tests/FSharp.Compiler.Service.Tests/FSharp.Compiler.Service.Tests.fsproj:5]: Yes something in Visual Studio is making this change automatically, I've seen it before in other branches.  I'll revert but it's a bit strange

File: tests/fsharp/tests.fs
Comment (dsyme) [reply tests/fsharp/tests.fs:1064]: Test names can't have `.` - it confuses the test explorer in VS

**Comment by KevinRansom** [comment]
I'll fix the test issues

**Comment by dsyme** [comment]
> I'll fix the test issues


@KevinRansom thanks!

**Comment by dsyme** [comment]
@KevinRansom all ready I think

**Review by KevinRansom — APPROVED** [review]

**Pull request merged**
